pkg/agent/ipassigner/responder/arp_responder.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package responder
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/mdlayher/arp"
+	"github.com/mdlayher/ethernet"
+	"k8s.io/klog/v2"
+)
+
+type arpResponder struct {
+	iface       *net.Interface
+	conn        *arp.Client
+	assignedIPs map[string]struct{}
+	closed      chan struct{}
+}
+
+var _ Responder = (*arpResponder)(nil)
+
+func NewARPResponder(iface *net.Interface) (*arpResponder, error) {
+	conn, err := arp.Dial(iface)
+	if err != nil {
+		return nil, fmt.Errorf("creating ARP responder for %q: %s", iface.Name, err)
+	}
+	return &arpResponder{
+		iface:       iface,
+		conn:        conn,
+		assignedIPs: make(map[string]struct{}),
+		closed:      make(chan struct{}),
+	}, nil
+}
+
+// GratuitousARP sends an gratuitous ARP packet for the IP.
+func (r *arpResponder) Gratuitous(ip net.IP) error {
+	select {
+	case <-r.closed:
+		return fmt.Errorf("ARP responder socket is closed")
+	default:
+		if ip.To4() == nil {
+			return fmt.Errorf("only IPv4 is supported")
+		}
+		pkt, err := arp.NewPacket(arp.OperationRequest, r.iface.HardwareAddr, ip, ethernet.Broadcast, ip)
+		if err != nil {
+			return err
+		}
+		return r.conn.WriteTo(pkt, ethernet.Broadcast)
+	}
+}
+
+func (r *arpResponder) InterfaceName() string {
+	return r.iface.Name
+}
+
+func (r *arpResponder) AssignIP(ip net.IP) error {
```

Comment:
Probably rename to "AddIP", "RemoveIP"? It is not really about assignment.

Reply:
update. thanks.